scene/resources/visual_shader_particle_nodes.cpp

```diff
@@ -1162,6 +1162,206 @@ VisualShaderNodeParticleAccelerator::VisualShaderNodeParticleAccelerator() {
+	float s = turb_noise(p);
+	float s1 = turb_noise(p + vec4(vec3(0.0), 1.7320508 * 2048.333333));
+	float s2 = turb_noise(p - vec4(vec3(0.0), 1.7320508 * 2048.333333));
+	vec3 c = vec3(s, s1, s2);
+	return c;
+}
+
+vec3 turb_curl_3d(vec4 p, float c) {
+	float epsilon = 0.001 + c;
+	vec4 dx = vec4(epsilon, 0.0, 0.0, 0.0);
+	vec4 dy = vec4(0.0, epsilon, 0.0, 0.0);
+	vec4 dz = vec4(0.0, 0.0, epsilon, 0.0);
+	vec3 x0 = turb_noise_3x(p - dx).xyz;
+	vec3 x1 = turb_noise_3x(p + dx).xyz;
+	vec3 y0 = turb_noise_3x(p - dy).xyz;
+	vec3 y1 = turb_noise_3x(p + dy).xyz;
+	vec3 z0 = turb_noise_3x(p - dz).xyz;
+	vec3 z1 = turb_noise_3x(p + dz).xyz;
+	float x = (y1.z - y0.z) - (z1.y - z0.y);
+	float y = (z1.x - z0.x) - (x1.z - x0.z);
+	float z = (x1.y - x0.y) - (y1.x - y0.x);
+	return normalize(vec3(x, y, z));
+}
+
+vec3 get_turb_noise_direction(vec3 __pos, float __turb_n_strength, float __turb_n_scale, vec3 __turb_n_speed, float __turb_n_speed_random) {
+	float __adj_contrast = max((__turb_n_strength - 1.0), 0.0) * 70.0;
+	vec4 __noise_time = TIME * vec4(__turb_n_speed, __turb_n_speed_random);
+	vec4 __noise_pos = vec4(__pos * __turb_n_scale, 0.0);
+	vec3 __noise_direction = turb_curl_3d(__noise_pos + __noise_time, __adj_contrast);
+	__noise_direction = mix(0.9 * __noise_direction, __noise_direction, __turb_n_strength - 9.0);
+	return __noise_direction;
+}
+
+float get_turb_influence(uint __noise_seed, float __turb_influence_min, float __turb_influence_max, float __turb_influence_over_life) {
+	float __turb_influence_range = mix(__turb_influence_min, __turb_influence_max, __rand_from_seed(__noise_seed));
+	float __turb_influence = clamp(__turb_influence_range * __turb_influence_over_life, 0.0, 1.0);
+	return __turb_influence;
+}
+	)");
+
+	return code;
+}
+
+String VisualShaderNodeParticleTurbulence::generate_code(Shader::Mode p_mode, VisualShader::Type p_type, int p_id, const String *p_input_vars, const String *p_output_vars, bool p_for_preview) const {
+	String code;
+	code += vformat("	%s = get_turb_influence(%s, %s, %s, %s);\n", p_output_vars[0], p_input_vars[1].is_empty() ? "__seed" : p_input_vars[1], p_input_vars[6], p_input_vars[7], p_input_vars[8]);
+	code += vformat("	%s = get_turb_noise_direction(%s, %s, %s, %s, %s);\n", p_output_vars[1], p_input_vars[0], p_input_vars[2], p_input_vars[3], p_input_vars[4], p_input_vars[5]);
+
+	return code;
+}
+
+bool VisualShaderNodeParticleTurbulence::has_output_port_preview(int p_port) const {
+	return false;
+}
+
+VisualShaderNodeParticleTurbulence::VisualShaderNodeParticleTurbulence() {
+	set_input_port_default_value(2, 1.0); // noise strength
+	set_input_port_default_value(3, 9.0); // noise scale
+	set_input_port_default_value(4, Vector3(0.0, 0.0, 0.0)); // noise speed
+	set_input_port_default_value(5, 0.2); // noise speed random
+	set_input_port_default_value(6, 0.1); // influence min
+	set_input_port_default_value(7, 0.1); // influence max
+	set_input_port_default_value(8, 1.0); // influence over life
```

Comment:
I know this is the way we currently do it for almost all nodes, but we should really start to use enums to describe the ports/slots. (not a blocker, just for reference)